src/gui/widgets/typed_debugger.cc

```diff
@@ -0,0 +1,894 @@
+                    if (linePartsCounter == 6) {
+                        break;
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+        switch (importType) {
+            case ImportType::DataTypes:
+                m_structs[name] = fields;
+                m_typeNames.push_back(name);
+                break;
+            case ImportType::Functions:
+                m_functions[address] = func;
+                break;
+            default:
+                break;
+        }
+    }
+
+    if (importType == ImportType::DataTypes) {
+        std::sort(m_typeNames.begin(), m_typeNames.end(), [](std::string left, std::string right) {
+            for (auto& letter : left) {
+                letter = tolower(letter);
+            }
+            for (auto& letter : right) {
+                letter = tolower(letter);
+            }
+            return left.compare(right) < 0;
+        });
+    }
+}
+
+PCSX::Widgets::TypedDebugger::TypedDebugger(bool& show) : m_show(show), m_listener(g_system->m_eventBus) {
+    m_listener.listen<PCSX::Events::ExecutionFlow::Reset>([this](const auto& event) {
+        m_displayedWatchData.clear();
+        for (const auto* bp : m_watchBreakpoints) {
+            g_emulator->m_debug->removeBreakpoint(bp);
+        }
+        m_watchBreakpoints.clear();
+        m_disabledInstructions.clear();
+
+        m_displayedFunctionData.clear();
+        for (const auto* bp : m_functionBreakpoints) {
+            g_emulator->m_debug->removeBreakpoint(bp);
+        }
+        m_functionBreakpoints.clear();
+        m_disabledFunctions.clear();
+    });
+}
+
+// Populates a node according to its type.
+void populate(WatchTreeNode* node,
+              std::unordered_map<std::string, PCSX::Widgets::TypedDebugger::StructFields>& structs_info) {
+    const auto type = node->type;
+
+    const std::regex arrayRegex(R"((.*)\[(\d+)\])");
+    std::smatch matches;
+    if (std::regex_match(type, matches, arrayRegex)) {
+        const auto type = matches[1].str();
+        const auto numChildren = std::stoul(matches[2].str());
+        for (size_t i = 0; i < numChildren; ++i) {
+            WatchTreeNode newNode{type, std::string(node->name + "[" + std::to_string(i) + "]"),
+                                  node->size / numChildren};
+            node->children.push_back(newNode);
+            populate(&node->children.back(), structs_info);
+        }
+    } else if (structs_info.contains(type)) {
+        const size_t numChildren = structs_info[type].size();
+
+        for (size_t i = 0; i < numChildren; ++i) {
+            const auto& field = structs_info[type][i];
+            WatchTreeNode newNode{field.type, field.name, field.size};
+            node->children.push_back(newNode);
+            populate(&node->children.back(), structs_info);
+        }
+    }
+}
+
+bool isInRAM(uint32_t address) {
```

Comment:
There's a lot of functions there that probably should be declared as `static` to avoid globals pollution.